src/Functions/generateSerialID.cpp

```diff
@@ -0,0 +1,167 @@
+#include "Common/Exception.h"
+#include <Common/ZooKeeper/ZooKeeper.h>
+#include <Columns/ColumnVector.h>
+#include <DataTypes/DataTypesNumber.h>
+#include <Functions/FunctionFactory.h>
+#include <Functions/FunctionHelpers.h>
+#include <Interpreters/Context.h>
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int KEEPER_EXCEPTION;
+}
+
+constexpr auto function_node_name = "/serial_ids/";
+constexpr size_t MAX_SERIES_NUMBER = 1000; // ?
+
+class FunctionSerial : public IFunction
+{
+private:
+    mutable zkutil::ZooKeeperPtr zk;
+    ContextPtr context;
+
+public:
+    static constexpr auto name = "generateSerialID";
+
+    explicit FunctionSerial(ContextPtr context_) : context(context_)
+    {
+        if (context->hasZooKeeper()) {
+            zk = context->getZooKeeper();
+        }
+    }
+
+    static FunctionPtr create(ContextPtr context)
+    {
+        return std::make_shared<FunctionSerial>(std::move(context));
+    }
+
+    String getName() const override { return name; }
+    size_t getNumberOfArguments() const override { return 1; }
+    bool isStateful() const override { return true; }
+    bool isDeterministic() const override { return false; }
+    bool isDeterministicInScopeOfQuery() const override { return false; }
+    bool isSuitableForConstantFolding() const override { return false; }
+    bool useDefaultImplementationForNulls() const override { return false; }
+    bool useDefaultImplementationForNothing() const override { return false; }
+    bool canBeExecutedOnDefaultArguments() const override { return false; }
+    bool isInjective(const ColumnsWithTypeAndName & /*sample_columns*/) const override { return true; }
+    bool hasInformationAboutMonotonicity() const override { return true; }
+    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return false; }
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        FunctionArgumentDescriptors mandatory_args{
+            {"series identifier", static_cast<FunctionArgumentDescriptor::TypeValidator>(&isStringOrFixedString), nullptr, "String or FixedString"}
+        };
+        validateFunctionArgumentTypes(*this, arguments, mandatory_args);
+
+        return std::make_shared<DataTypeInt64>();
+    }
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t input_rows_count) const override
+    {
+        if (zk == nullptr)
+            throw Exception(ErrorCodes::KEEPER_EXCEPTION,
+            "ZooKeeper is not configured for function {}",
+            getName());
+        if (zk->expired())
+            zk = context->getZooKeeper();
+
+        // slow?
+        if (zk->exists(function_node_name) && zk->getChildren(function_node_name).size() == MAX_SERIES_NUMBER) {
```

Comment:
Instead of `getChildren()` you can use `exists(...)`  optional `stats` out parameter that has `numChildren` field.